File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/HMSTrackSettingsExtension.kt
package live.hms.hmssdk_flutter

import live.hms.video.media.settings.HMSAudioTrackSettings
import live.hms.video.media.settings.HMSTrackSettings
import live.hms.video.media.settings.HMSVideoTrackSettings
import live.hms.video.media.settings.PhoneCallState
import live.hms.video.sdk.HMSSDK

class HMSTrackSettingsExtension {
    companion object {
        fun toDictionary(hmssdk: HMSSDK): HashMap<String, Any>? {
            val map = HashMap<String, Any>()
            val hmsTrackSettings: HMSTrackSettings = hmssdk.hmsSettings

            if (hmsTrackSettings.videoSettings != null) {
                map["video_track_setting"] = HMSVideoTrackSettingsExtension.toDictionary(hmsTrackSettings.videoSettings)!!
            }

            if (hmsTrackSettings.audioSettings != null) {
                map["audio_track_setting"] = HMSAudioTrackSettingsExtension.toDictionary(hmsTrackSettings.audioSettings)!!
            }

            return map
        }

        fun setTrackSettings(
            hmsAudioTrackHashMap: HashMap<String, Any?>?,
            hmsVideoTrackHashMap: HashMap<String, Any?>?,
        ): HMSTrackSettings {
            var hmsAudioTrackSettings = HMSAudioTrackSettings.Builder()

            hmsAudioTrackSettings.setUseHardwareAcousticEchoCanceler(false)
            hmsAudioTrackSettings.setPhoneCallMuteState(PhoneCallState.DISABLE_MUTE_ON_VOIP_PHONE_CALL_RING)

            hmsAudioTrackHashMap?.let { audioHashMap ->

                val initialState = HMSTrackInitStateExtension.getHMSTrackInitStatefromValue(audioHashMap["track_initial_state"] as String)
                hmsAudioTrackSettings = hmsAudioTrackSettings.initialState(initialState)

                val useHardwareAcousticEchoCanceler =
                    audioHashMap["user_hardware_acoustic_echo_canceler"] as Boolean?
                useHardwareAcousticEchoCanceler?.let { useHardware ->
                    hmsAudioTrackSettings.setUseHardwareAcousticEchoCanceler(useHardware)
                }

                val phoneCallMuteState = audioHashMap["phone_call_mute_state"] as? String
                phoneCallMuteState?.let { callMuteState ->
                    when (callMuteState) {
                        "ENABLE_MUTE_ON_PHONE_CALL_RING" ->
                            hmsAudioTrackSettings.setPhoneCallMuteState(
                                PhoneCallState.ENABLE_MUTE_ON_PHONE_CALL_RING,
                            )
                        else -> hmsAudioTrackSettings.setPhoneCallMuteState(PhoneCallState.DISABLE_MUTE_ON_VOIP_PHONE_CALL_RING)
                    }
                }
            }

            var hmsVideoTrackSettings = HMSVideoTrackSettings.Builder()
            if (hmsVideoTrackHashMap != null) {
                val cameraFacing = getHMSCameraFacingFromValue(hmsVideoTrackHashMap["camera_facing"] as String?)
                val disableAutoResize = hmsVideoTrackHashMap["disable_auto_resize"] as Boolean
                val initialState =
                    HMSTrackInitStateExtension.getHMSTrackInitStatefromValue(
                        hmsVideoTrackHashMap["track_initial_state"] as String,
                    )
                val forceSoftwareDecoder = hmsVideoTrackHashMap["force_software_decoder"] as Boolean

                if (cameraFacing != null) {
                    hmsVideoTrackSettings = hmsVideoTrackSettings.cameraFacing(cameraFacing)
                }
                if (disableAutoResize != null) {
                    hmsVideoTrackSettings = hmsVideoTrackSettings.disableAutoResize(disableAutoResize)
                }
                if (initialState != null) {
                    hmsVideoTrackSettings = hmsVideoTrackSettings.initialState(initialState)
                }
                if (forceSoftwareDecoder != null) {
                    hmsVideoTrackSettings = hmsVideoTrackSettings.forceSoftwareDecoder(forceSoftwareDecoder)
                }
            }

            return HMSTrackSettings.Builder().audio(hmsAudioTrackSettings.build()).video(hmsVideoTrackSettings.build()).build()
        }

        private fun getHMSCameraFacingFromValue(cameraFacing: String?): HMSVideoTrackSettings.CameraFacing {
            return when (cameraFacing) {
                "back" -> HMSVideoTrackSettings.CameraFacing.BACK
                "front" -> HMSVideoTrackSettings.CameraFacing.FRONT
                else -> HMSVideoTrackSettings.CameraFacing.FRONT
            }
        }
    }
}

File: packages/hmssdk_flutter/ios/Classes/Models/HMSTrackSettingsExtension.swift
//
//  HMSTrackSettingsExtension.swift
//  hmssdk_flutter
//
//  Created by Govind on 04/08/22.
//

import Foundation
import HMSSDK

class HMSTrackSettingsExtension {

    static var videoSettings: HMSVideoTrackSettings?

    static func toDictionary(_ hmssdk: HMSSDK, _ audioMixerSourceMap: [String: HMSAudioNode]?) -> [String: Any] {

        let hmsTrackSettings = hmssdk.trackSettings
        var dict = [String: Any]()

        if let hmsVideoSettings = hmsTrackSettings.video {
            dict["video_track_setting"] = HMSTrackSettingsExtension.toDictionary(hmsVideoSettings)
        }
        if let hmsAudioSettings = hmsTrackSettings.audio {
            dict["audio_track_setting"] = HMSTrackSettingsExtension.toDictionary(hmsAudioSettings, audioMixerSourceMap!)
        }

        return dict
    }

    static func toDictionary(_ hmsVideoTrackSettings: HMSVideoTrackSettings) -> [String: Any] {

        var dict = [String: Any]()
        if hmsVideoTrackSettings.codec ==  HMSCodec.H264 {
            dict["video_codec"] = "h264"
        } else {
            dict["video_codec"] = "vp8"
        }
        if hmsVideoTrackSettings.cameraFacing == .front {
            dict["camera_facing"] = "front"
        } else {
            dict["camera_facing"] = "back"
        }
        dict["max_bitrate"] = hmsVideoTrackSettings.maxBitrate
        dict["max_frame_rate"] = hmsVideoTrackSettings.maxFrameRate
        dict["resolution"] = HMSVideoResolutionExtension.toDictionary(hmsVideoTrackSettings.resolution)
        dict["simulcast_settings"] = hmsVideoTrackSettings.simulcastSettings
        dict["track_description"] = hmsVideoTrackSettings.trackDescription
        dict["video_plugins"] = hmsVideoTrackSettings.videoPlugins
        return dict
    }

    static func toDictionary(_ hmsAudioTrackSettings: HMSAudioTrackSettings, _ audioMixerSourceMap: [String: HMSAudioNode]) -> [String: Any] {

        var dict = [String: Any]()

        dict["track_description"] = hmsAudioTrackSettings.trackDescription
        dict["max_bitrate"] = hmsAudioTrackSettings.maxBitrate
        dict["audio_source"] = audioMixerSourceMap.keys.map {$0}
        dict["audio_mode"] = getStringFromAudioMode(from: hmsAudioTrackSettings.audioMode)
        return dict
    }

    static func setTrackSetting(_ settingsDict: [AnyHashable: Any],
                                _ audioMixerSourceMap: [String: HMSAudioNode],
                                _ vbConformer: HMSVirtualBackgroundActionPluginProtocol,
                                _ result: @escaping FlutterResult) -> HMSTrackSettings? {

        var audioSettings: HMSAudioTrackSettings?

        if let audioSettingsDict = settingsDict["audio_track_setting"] as? [AnyHashable: Any] {

            var initialMuteState: HMSTrackMuteState?
            if let muteState = audioSettingsDict["track_initial_state"] as? String {
                initialMuteState = getinitialMuteState(from: muteState)
            }

            if #available(iOS 13.0, *), !audioMixerSourceMap.isEmpty {
                do {
                    let audioMixerSource = try HMSAudioMixerSource(nodes: audioMixerSourceMap.values.map {$0})

                    audioSettings = HMSAudioTrackSettings.build({ builder in

                        builder.audioSource = audioMixerSource

                        if let muteState = initialMuteState {
                            builder.initialMuteState = muteState
                        }

                        if let mode = getAudioMode(from: audioSettingsDict["audio_mode"] as? String) {
                            builder.audioMode = mode
                        }

                        /*
                         Here we set the noise cancellation controller based on the parameter passed in audio track
                         settings
                         */
                        if let enableNoiseCancellation = audioSettingsDict["enable_noise_cancellation"] as? Bool {
                            if enableNoiseCancellation {

                                /// We create  the noise cancellation plugin
                                HMSNoiseCancellationController.createPlugin()
                                builder.noiseCancellationPlugin = HMSNoiseCancellationController.noiseCancellationController
                            }
                        }
                    })

                } catch {
                    print(#function, HMSErrorExtension.toDictionary(error))
                    result(false)
                }
            } else {
                audioSettings = HMSAudioTrackSettings.build({ builder in

                    if let muteState = initialMuteState {
                        builder.initialMuteState = muteState
                    }

                    if let mode = getAudioMode(from: audioSettingsDict["audio_mode"] as? String) {
                        builder.audioMode = mode
                    }

                    /*
                     Here we set the noise cancellation controller based on the parameter passed in audio track
                     settings
                     */
                    if let enableNoiseCancellation = audioSettingsDict["enable_noise_cancellation"] as? Bool {
                        if enableNoiseCancellation {

                            /// We create  the noise cancellation plugin
                            HMSNoiseCancellationController.createPlugin()
                            builder.noiseCancellationPlugin = HMSNoiseCancellationController.noiseCancellationController
                        }
                    }

                })
            }
        }

        if let videoSettingsDict = settingsDict["video_track_setting"] as? [AnyHashable: Any] {
            if let cameraFacing = videoSettingsDict["camera_facing"] as? String,
               let isVirtualBackgroundEnabled = videoSettingsDict["is_virtual_background_enabled"] as? Bool,
               let initialMuteState = videoSettingsDict["track_initial_state"] as? String {

                var videoPlugins: [HMSVideoPlugin]?
                if isVirtualBackgroundEnabled {
                    if #available(iOS 15.0, *) {
                        if let virtualbackground = vbConformer.plugin {
                            videoPlugins = []
                            videoPlugins?.append(virtualbackground)
                        }
                    } else {
                        HMSErrorLogger.logError("\(#function)", "Virtual Background is not supported below iOS 15", "Plugin not supported error")
                    }
                }
                videoSettings = HMSVideoTrackSettings(codec: HMSCodec.VP8,
                                                      resolution: .init(width: 320, height: 180),
                                                      maxBitrate: 32,
                                                      maxFrameRate: 30,
                                                      cameraFacing: getCameraFacing(from: cameraFacing),
                                                      simulcastSettings: nil,
                                                      trackDescription: "track_description",
                                                      initialMuteState: getinitialMuteState(from: initialMuteState),
                                                      videoPlugins: videoPlugins)
            }
        }

        return HMSTrackSettings(videoSettings: videoSettings, audioSettings: audioSettings)
    }

    static private func getCameraFacing(from string: String) -> HMSCameraFacing {
        if string.lowercased().contains("back") {
            return HMSCameraFacing.back
        }
        return HMSCameraFacing.front
    }

    static private func getinitialMuteState(from string: String) -> HMSTrackMuteState {
        if string.lowercased().contains("unmuted") {
            return HMSTrackMuteState.unmute
        }
        return HMSTrackMuteState.mute
    }

    static private  func getAudioMode(from mode: String?) -> HMSAudioMode? {
        switch mode {
        case "voice":
            return .voice

        case "music":
            return .music

        default:
            return nil
        }
    }

    static private func getStringFromAudioMode(from mode: HMSAudioMode) -> String? {
        switch mode {
        case .music:
            return "music"

        case .voice:
            return "voice"

        default:
            return nil
        }
    }
}
